File: packages/brmem/src/brmem/list.py
"""List Branch Memory Entries."""

from __future__ import annotations

from pathlib import Path

import click

from asdl_core.clinkr.context import load_typed_context
from asdl_core.clinkr.ensure import Ensure
from asdl_core.clinkr.exit import ClinkrExit
from asdl_core.clinkr.models import ClinkrModel
from asdl_core.clinkr.operation import clinkr_operation
from brmem.context import BrmemCliContext
from brmem.key_validation import check_key
from brmem.ref_layout import (
    BASE_NAMESPACE,
    EntryRef,
    check_branch_name,
    check_namespace,
    namespace_display_label,
    normalize_namespace_option,
)
from brmem.validation import first_failure

_ALL_NAMESPACES_SCOPE = "all"


class ListEntriesRequest(ClinkrModel):
    namespace: str | None = None
    key: str | None = None
    branch: str | None = None
    base: bool = False
    all_branches: bool = False


class ListEntriesResult(ClinkrModel):
    namespace_scope: str
    key: str | None
    branch: str
    base: bool
    all_branches: bool
    entries: list[EntryRef]


def render_list_entries(result: ListEntriesResult) -> None:
    for entry in result.entries:
        click.echo(
            f"{namespace_display_label(entry.namespace)} | "
            f"Entry Key {entry.key} | Branch {entry.branch}"
        )


@clinkr_operation(
    name="list",
    help=(
        "List Branch Memory Entries. Defaults to the current branch; "
        "pass --branch to override or --all-branches to include every branch. "
        "--namespace and --key further filter. Pass --base to restrict to ad-hoc base Entries."
    ),
    human_renderer=render_list_entries,
)
def run_list_entries(
    ctx: click.Context,
    request: ListEntriesRequest,
) -> ClinkrExit[ListEntriesResult]:
    brmem_context = load_typed_context(ctx, BrmemCliContext)

    Ensure.true(
        not (request.base and request.namespace is not None),
        error_type="base_and_namespace_conflict",
        message="--base and --namespace are mutually exclusive.",
    )
    Ensure.true(
        not (request.branch is not None and request.all_branches),
        error_type="branch_and_all_branches_conflict",
        message="--branch and --all-branches are mutually exclusive.",
    )

    all_namespaces = not request.base and request.namespace is None
    scope_namespace = _scope_namespace(request)
    validation_failure = first_failure(
        (
            "invalid_namespace",
            None if all_namespaces else check_namespace(scope_namespace),
        ),
        ("invalid_key", None if request.key is None else check_key(request.key)),
        (
            "invalid_branch_name",
            None if request.branch is None else check_branch_name(request.branch),
        ),
    )
    error_type, message = validation_failure or ("", "")
    Ensure.true(
        validation_failure is None,
        error_type=error_type,
        message=message,
    )

    branch = None
    if not request.all_branches:
        branch = (
            request.branch
            if request.branch is not None
            else Ensure.ideal_state(brmem_context.git_gateway.get_current_branch(Path.cwd()))
        )

    if all_namespaces:
        entries = brmem_context.brmem_gateway.list_all_entries(
            key=request.key,
            branch=branch,
        )
        namespace_scope = _ALL_NAMESPACES_SCOPE
    else:
        entries = brmem_context.brmem_gateway.list_entries(
            namespace=scope_namespace,
            key=request.key,
            branch=branch,
        )
        namespace_scope = scope_namespace

    return ClinkrExit.ok(
        ListEntriesResult(
            namespace_scope=namespace_scope,
            key=request.key,
            branch=branch,
            base=request.base,
            all_branches=request.all_branches,
            entries=entries,
        )
    )


def _scope_namespace(request: ListEntriesRequest) -> str:
    if request.base:
        return BASE_NAMESPACE
    return normalize_namespace_option(request.namespace)

File: ts/packages/brmem/src/operations/list.ts
import { failure, ok, type RenderCapabilities } from "@asdl/clinkr";
import { renderTextTable } from "@asdl/core/text-table";
import { z } from "zod";

import type { BrmemCliContext } from "../context.ts";
import {
	BASE_NAMESPACE,
	namespaceDisplayLabel,
	normalizeNamespaceOption,
	type EntryRef,
} from "../ref-layout.ts";
import {
	firstFailure,
	validateBranchName,
	validateEntryKey,
	validateNamespaceName,
	validationMessage,
} from "../validation.ts";
import { gatewayFailure, resolveCurrentBranch } from "./shared.ts";

const ALL_NAMESPACES_SCOPE = "all";

export const listRequestSchema = z.object({
	namespace: z.string().optional().describe("Namespace filter. Omit for all Namespaces."),
	key: z.string().optional().describe("Exact Entry Key filter."),
	branch: z.string().optional().describe("Branch. Defaults to current branch."),
	base: z.boolean().default(false).describe("Restrict to Base Namespace."),
	allBranches: z.boolean().default(false).describe("List Entries from all branches."),
});

export const listResultSchema = z.object({
	namespace_scope: z.string(),
	key: z.string().nullable(),
	branch: z.string().nullable(),
	base: z.boolean(),
	all_branches: z.boolean(),
	entries: z.array(
		z.object({
			namespace: z.string(),
			key: z.string(),
			branch: z.string(),
			ref_name: z.string(),
		}),
	),
});

export type ListRequest = z.infer<typeof listRequestSchema>;
export type ListResult = z.infer<typeof listResultSchema>;

export async function runList(ctx: BrmemCliContext, request: ListRequest) {
	if (request.base && request.namespace !== undefined) {
		return failure("base_and_namespace_conflict", "--base and --namespace are mutually exclusive.");
	}
	if (request.branch !== undefined && request.allBranches) {
		return failure(
			"branch_and_all_branches_conflict",
			"--branch and --all-branches are mutually exclusive.",
		);
	}
	const shouldListAllNamespaces = !request.base && request.namespace === undefined;
	const scopeNamespace = request.base
		? BASE_NAMESPACE
		: normalizeNamespaceOption(request.namespace);
	const validationFailure = firstFailure(
		[
			"invalid_namespace",
			shouldListAllNamespaces
				? undefined
				: validationMessage("namespace", scopeNamespace, validateNamespaceName(scopeNamespace)),
		],
		[
			"invalid_key",
			request.key === undefined
				? undefined
				: validationMessage("key", request.key, validateEntryKey(request.key)),
		],
		[
			"invalid_branch_name",
			request.branch === undefined
				? undefined
				: validationMessage("branch name", request.branch, validateBranchName(request.branch)),
		],
	);
	if (validationFailure !== undefined) return failure(validationFailure[0], validationFailure[1]);
	let branch: string | undefined;
	if (!request.allBranches) {
		if (request.branch !== undefined) branch = request.branch;
		else {
			const resolvedBranch = await resolveCurrentBranch(ctx);
			if (typeof resolvedBranch !== "string") return resolvedBranch;
			branch = resolvedBranch;
		}
	}
	const entriesResult = shouldListAllNamespaces
		? await ctx.gateway.listAllEntries({ key: request.key, branch })
		: await ctx.gateway.listEntries({ namespace: scopeNamespace, key: request.key, branch });
	if (entriesResult.type === "error") return gatewayFailure<ListResult>(entriesResult.error);
	return ok({
		namespace_scope: shouldListAllNamespaces ? ALL_NAMESPACES_SCOPE : scopeNamespace,
		key: request.key ?? null,
		branch: branch ?? null,
		base: request.base,
		all_branches: request.allBranches,
		entries: entriesResult.value.map(entryJson),
	});
}

export function renderList(
	result: ListResult,
	caps: RenderCapabilities = { canEmitAnsi: false },
): string {
	if (result.entries.length === 0) return "";
	return renderTextTable({
		columns: [
			{ header: "NAMESPACE", style: "bold-cyan" },
			{ header: "ENTRY KEY" },
			{ header: "BRANCH" },
		],
		rows: result.entries.map((entry) => [
			namespaceDisplayLabel(entry.namespace),
			entry.key,
			entry.branch,
		]),
		canEmitAnsi: caps.canEmitAnsi,
		shouldDrawRule: true,
		headerStyle: "bold-cyan",
	});
}

function entryJson(entry: EntryRef): {
	namespace: string;
	key: string;
	branch: string;
	ref_name: string;
} {
	return {
		namespace: entry.namespace,
		key: entry.key,
		branch: entry.branch,
		ref_name: entry.entryLocator,
	};
}
